### PR TITLE
177 183 inputs as attributes

### PR DESCRIPTION
**Pull request by t-sanchez**
Fix the issue https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/issues/177 -> Cleaned up inputs to workflows.
Fix https://github.com/Medical-Image-Analysis-Laboratory/mialsuperresolutiontoolkit/issues/183 -> Make the pipeline crash explicitly when

A single stack is run
A stack isn't found in filterbyrunid

### CODE REVIEW

**Review by pdedumast — COMMENTED** [review]
Is the case where "stacks" is not specified in the JSON file but only one single stack is available in the subject's exam managed?

**Comment by t-sanchez** [comment]
That's actually a good point.
I don't think that it's currently handled. I'd need to check this case specifically.

**Comment by t-sanchez** [comment]
Actually, rather than having the checks scattered throughout the input stage, I refactored `FilteringByRunId` into `CheckAndFilterInputStacks`. The class takes as input `input_images` (T2w images), and optionally stacks, masks and labels. 

It either uses the given stacks or extracts them from the T2w images and then checks that masks and labels have the requested stacks. It raises an exception when it is not the case. It also checks that at least one stack is requested/found.

I think that this simplifies the workflow in the input_stage, removing the need for individual filtering. Rather, we filter immediately the files based on the stacks or the input_images found. The stack ordering then works on already filtered data, and there is no need to re-filter them afterwards.

**Comment by t-sanchez** [comment]
As the black refactoring makes the reading of relevant changes quite difficult, I've put some comments around the important changes.

File: pymialsrtk/interfaces/postprocess.py [hunk 1]
```diff
@@ -32,29 +39,50 @@
 #  Refinement HR mask
 #######################
 
+
 class MialsrtkRefineHRMaskByIntersectionInputSpec(BaseInterfaceInputSpec):
```
Comment (t-sanchez) [inline pymialsrtk/interfaces/postprocess.py:43]: Put back attributes as inputs `MialsrtkRefineHRMaskByIntersection`.

File: pymialsrtk/interfaces/postprocess.py [hunk 2]
```diff
@@ -216,279 +263,392 @@ def _list_outputs(self):
...
 ############################
 
+
 class FilenamesGenerationInputSpec(BaseInterfaceInputSpec):
     """Class used to represent inputs of the FilenamesGeneration interface."""
 
-    stacks_order = traits.List(mandatory=True,
-                               desc='List of stack run-id that specify the'
-                                    ' order of the stacks')
+    sub_ses = traits.Str(
```
Comment (t-sanchez) [inline pymialsrtk/interfaces/postprocess.py:270]: Put back attributes as inputs `FilenamesGeneration`.

File: pymialsrtk/interfaces/preprocess.py [hunk 3]
```diff
@@ -509,110 +589,173 @@ class MialsrtkMaskImage(BaseInterface):
...
 
-
 ###############################
 # Stacks ordering and filtering
 ###############################
 
 
-class FilteringByRunidInputSpec(BaseInterfaceInputSpec):
-    """Class used to represent inputs of the FilteringByRunid interface."""
+class CheckAndFilterInputStacksInputSpec(BaseInterfaceInputSpec):
```
Comment (t-sanchez) [inline pymialsrtk/interfaces/preprocess.py:623]: Replace the class `FilteringByRunid` by `CheckAndFilterInputStacks`
The point is to have a single class that deals with all inputs and filters it based on either 1. given stacks inputs or 2. available stacks in input_images.

Comment (t-sanchez) [reply pymialsrtk/interfaces/preprocess.py:623]: As files are already filtered, there's no need for additional filtering down the line.

File: pymialsrtk/workflows/input_stage.py [hunk 4]
```diff
@@ -64,245 +65,209 @@ def create_input_stage(p_bids_dir,
...
-            custom_masks_filter.inputs.stacks_id = p_stacks
+            interface=IdentityInterface(fields=["out_file"]),
+            name="brain_masks_bypass",
+            iterfield=["out_file"],
+        )
 
     else:
-
-        if p_stacks is not None:
-            t2ws_filter_prior_masks = pe.Node(
```
Comment (t-sanchez) [inline pymialsrtk/workflows/input_stage.py:161]: All filtering is removed as it's centralized in check_input

File: pymialsrtk/workflows/input_stage.py [hunk 5]
```diff
@@ -64,245 +65,209 @@ def create_input_stage(p_bids_dir,
...
+    dg_fields = ["T2ws"]
     if p_use_manual_masks:
-        dg_fields += ['masks']
+        dg_fields += ["masks"]
     if p_do_reconstruct_labels:
-        dg_fields += ['labels']
+        dg_fields += ["labels"]
 
     dg = pe.Node(
         interface=DataGrabber(
```
Comment (t-sanchez) [inline pymialsrtk/workflows/input_stage.py:95]: In various places, move some inline attributes declaration to the function level.

File: pymialsrtk/workflows/input_stage.py [hunk 6]
```diff
@@ -64,245 +65,209 @@ def create_input_stage(p_bids_dir,
...
+                "Unet.ckpt-20000.index",
+            ),
+        ).split(".index")[0]
         brainMask.inputs.threshold_seg = 0.5
 
-    t2ws_filtered = pe.Node(interface=preprocess.FilteringByRunid(),
-                            name='t2ws_filtered')
-    masks_filtered = pe.Node(interface=preprocess.FilteringByRunid(),
-                             name='masks_filtered')
+    check_input = pe.Node(
```
Comment (t-sanchez) [inline pymialsrtk/workflows/input_stage.py:186]: Main addition: check_input to run between data grabber and brain extraction / mask utility bypassing.

File: pymialsrtk/workflows/input_stage.py [hunk 7]
```diff
@@ -64,245 +65,209 @@ def create_input_stage(p_bids_dir,
...
+            interface=IdentityInterface(fields=["stacks_order"]),
+            name="stackOrdering",
+        )
         stacksOrdering.inputs.stacks_order = p_stacks
-        stacksOrdering.inputs.verbose = p_verbose
 
     if p_do_srr_assessment:
 
         rg = pe.Node(
             interface=DataGrabber(
```
Comment (t-sanchez) [inline pymialsrtk/workflows/input_stage.py:210]: Add fields as attributed.

File: pymialsrtk/workflows/postproc_stage.py [hunk 8]
```diff
@@ -71,170 +71,218 @@ def create_postproc_stage(
...
-    srtkMaskImage02.inputs.verbose = p_verbose
+        interface=preprocess.MialsrtkMaskImage(verbose=p_verbose),
+        name="srtkMaskImage02",
+    )
 
     if p_do_anat_orientation and p_ga is not None:
 
         ga_str = convert_ga(p_ga)
 
         atlas_grabber = pe.Node(
```
Comment (t-sanchez) [inline pymialsrtk/workflows/postproc_stage.py:108]: Move some fields directly in the datagrabber definition

File: pymialsrtk/workflows/preproc_stage.py [hunk 9]
```diff
@@ -67,240 +67,336 @@ def create_preproc_stage(
...
+    )
 
     srtkCorrectSliceIntensity01 = pe.MapNode(
-        interface=preprocess.MialsrtkCorrectSliceIntensity(),
-        name='srtkCorrectSliceIntensity01',
-        iterfield=['in_file', 'in_mask'])
-    srtkCorrectSliceIntensity01.inputs.out_postfix = '_uni'
-    srtkCorrectSliceIntensity01.inputs.verbose = p_verbose
-
+        interface=preprocess.MialsrtkCorrectSliceIntensity(
```
Comment (t-sanchez) [inline pymialsrtk/workflows/preproc_stage.py:115]: Move verbose directly to node declaration.

**Review by pdedumast — APPROVED** [review]

**Pull request merged**